### PR TITLE
Re-add PlayerGiveXpEvent & PlayerChangeLevelEvent

### PR DESCRIPTION
**Pull request by FireController1847**
PR #3023 added these a while ago, but they appear to have been lost after 1.12.2. This re-adds them

It also reorganizes the PlayerPickupXpEvent. The events go like so:

Similar to the `PlayerEvent` class, all experience events are in a `PlayerXpEvent` class. The events will be called `PlayerXpEvent#PickupXp`, `PlayerXpEvent#XpChange`, and `PlayerXpEvent#LevelChange`. The `PlayerPickupXpEvent` will become deprecated but will still work, and will be extended from `PlayerXpEvent#PickupXp`.

### CODE REVIEW

**Comment by srs-bsns** [comment]
Perhaps those event classes should be made static inner classes of `PlayerEvent` like the rest of the PlayerEvent subclasses.

**Comment by FireController1847** [comment]
I thought about that, but I wanted to keep all of the XP-related classes in the same location. The already-made PlayerPickupXpEvent was made as its own class, so that's where I put the others

File: patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch [hunk 1]
```diff
@@ -425,7 +425,29 @@
...
     }
  
-@@ -1703,7 +1818,10 @@
+@@ -1462,6 +1577,10 @@
+    }
+ 
+    public void func_195068_e(int p_195068_1_) {
++      net.minecraftforge.event.entity.player.PlayerGiveXpEvent event = new net.minecraftforge.event.entity.player.PlayerGiveXpEvent(this, p_195068_1_);
++      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
++      p_195068_1_ = event.getAmount();
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch:434]: Why? The amount is never changed in the event (why not though?).

**Review by diesieben07 — CHANGES_REQUESTED** [review]
Is there any way we could get this into one event? They are very similar in purpose after all.

File: patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch (hunk 1 above)
Comment (FireController1847) [reply patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch:434]: Whoops, I forgot to implement that. I'll add a setter to it

**Comment by FireController1847** [comment]
@diesieben07 Sure, but how so? The events are similar but not the same, they are called at different points in time for different things. For example, the XpPickupEvent is called when the player touches experience, the XpGiveEvent is called when experience is added to the player (which can and will ultimately be called by the XpPickupEvent), and the ChangeLevelEvent is called when the Xp gets high enough to switch to another level (which can be called by the XpGiveEvent which can result from an XpPickupEvent).

These are all different methods within the different classes. The EntityPlayer has two main experience methods: #giveExperiencePoints and #addExperienceLevel. Admittedly, one is called by the other, but not always. Similarly, as you probably know, the event for XpPickupEvent is called in the ExperienceOrbEntity#onCollideWithPlayer method.

Due to each of these being their own method, they'd all have their own use-cases for cancelling and disabling. The XpPickupEvent could be cancelled to summon a different type of orb or used to change the value of the orb. The GiveXpEvent could be cancelled to allow a modder to use their own experience event and still have it respond to commands (something strongly lacking from the XpPickupEvent). The ChangeLevelEvent could be used to help the modder monitor whether or not the player changed their level to do something else, without having to implement &monitor it themselves by basing it on the GiveXpEvent and/or the PlayerTick event.

I could probably put all three events into a "PlayerXpEvent", kind of like the "PlayerEvent" class. Would that work? Or did you have something else in mind?

Also, if I were to make the overarching "PlayerXpEvent" class, would we also want to move the XpPickupEvent to said class? That would be a breaking change, though.

**Comment by diesieben07** [comment]
It was just an idea, if it's not feasible, so be it :D

**Comment by gigaherz** [comment]
I think this is a good chance to make the name of the event PlayerGainXpEvent, since it apparently gets called any time the player gets xp added to its counter.

Although, I also think it would be nice to move these as sub-events of PlayerEvent itself. 

Ie. PlayerEvent.PickupXp / PlayerEvent.GainXp / PlayerEvent.ChangeLevel

For backward compatibility purposes, PuckupXp could be temporarily defined as extending PlayerPickupXpEvent, which already extends PlayerEvent.

**Comment by FireController1847** [comment]
@gigaherz I made a prototype recently that made a `PlayerXpEvent`, and it moved all three into it. Then you could use it like `PlayerXpEvent#PickupXp`, `PlayerXpEvent#GiveXp`, and `PlayerXpEvent#ChangeLevel.` I disagree with calling it `GainXp` because the event is also fired when the player loses xp. Admittedly `GiveXp` is not much better in that regard, however I chose that because that's what the method is called in `PlayerEntity`. Any ideas on a name that would apply to both losing and gaining experience? Maybe `PlayerXpEvent#ChangeXp`, similar to `ChangeLevel`. Both of these don't really run off the tongue though.

**Comment by gigaherz** [comment]
If it also fires when losing XP, I'd call it XpChanged, similarly I'd call the other one LevelChanged

**Comment by FireController1847** [comment]
@gigaherz I like that, however it's fired before it is technically changed and it allows you to edit it before it goes through, so past tense wouldn't really work. Might as well just call it XpChange and LevelChange. Do you agree with the `PlayerXpEvents` class idea?

**Comment by gigaherz** [comment]
I don't mind either way. Both adding it to PlayerEvent, or having a dedicated PlayerXpEvents class sound fine to me. And yes, since it's called before, it would have to be XpChanging, but I agree, XpChange sounds better.

**Comment by FireController1847** [comment]
Okay, this has been tested and is ready to merge. All events have been added and the required events have been deprecated.

**Review by diesieben07 — CHANGES_REQUESTED** [review]
Unfortunately your extends/double-firing logic is flawed. People subscribing to `PlayerXpEvent.PickupXp` will receive _both_ events you are firing. You should only fire the old one, that way subscribers to old and new will receive it.

**Comment by FireController1847** [comment]
@diesieben07 I tried that at first and for some reason it didn't fire the newer event. I'll do some more testing today to see if I can get it working the proper way. It's kind of difficult to test because for some reason every time I get an event on my test mod it fires like 5-10 times, spamming my console. I figured that didn't have to do with any of my changes, it just happened all the sudden.

EDIT: I did some more testing and now it's working by only firing the old one. I think my previous testing was calling the new one and expecting the old one to fire :eyes:. In terms of the multiple-calling, it's still doing it, but I think it's just an issue with the build process. Now each even is being consistently called 4 times, instead of the 4 & 8 times before (I should have noticed it was firing twice). I'll push my fix now, _then_ it should be ready to merge :)

**Review by diesieben07 — APPROVED** [review]

**Comment by FireController1847** [comment]
Looking forward to when this gets merged as I can't complete the rewrite of my mod without it :)

File: patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch [hunk 2]
```diff
@@ -425,7 +425,29 @@
...
     }
  
-@@ -1703,7 +1818,10 @@
+@@ -1462,6 +1577,10 @@
+    }
+ 
+    public void func_195068_e(int p_195068_1_) {
++      net.minecraftforge.event.entity.player.PlayerXpEvent.XpChange event = new net.minecraftforge.event.entity.player.PlayerXpEvent.XpChange(this, p_195068_1_);
++      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
++      p_195068_1_ = event.getAmount();
```
Comment (DaemonUmbra) [inline patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch:434]: This should be moved to ForgeEventFactory

File: patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch [hunk 3]
```diff
@@ -425,7 +425,29 @@
...
+       this.func_85039_t(p_195068_1_);
+       this.field_71106_cc += (float)p_195068_1_ / (float)this.func_71050_bK();
+       this.field_71067_cb = MathHelper.func_76125_a(this.field_71067_cb + p_195068_1_, 0, Integer.MAX_VALUE);
+@@ -1501,6 +1620,10 @@
+    }
+ 
+    public void func_82242_a(int p_82242_1_) {
++      net.minecraftforge.event.entity.player.PlayerXpEvent.LevelChange event = new net.minecraftforge.event.entity.player.PlayerXpEvent.LevelChange(this, p_82242_1_);
++      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
++      p_82242_1_ = event.getLevels();
```
Comment (DaemonUmbra) [inline patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch:445]: This too

**Review by DaemonUmbra — CHANGES_REQUESTED** [review]
This is the distinction I got from Lex regarding where things go:

> if they are small patches, inline, if they are just firing an event, [ForgeEvent]Factory, if they are large logic blocks, [Forge]Hooks

[Source on Discord](https://discordapp.com/channels/313125603924639766/313125603924639766/623049613427081226)

**Comment by FireController1847** [comment]
I'm not trying to be _too_ pushy, but as far as I know the current implementation follows the standards of the rest of forge and is working entirely. I cannot finish updating my mod without this.

@DaemonUmbra Have you talked with LexManos/CPW about that change? Do we still want me to make the change or no?

**Comment by DaemonUmbra** [comment]
https://discordapp.com/channels/313125603924639766/313125603924639766/627404624516874260

**Comment by FireController1847** [comment]
I'm not entirely sure what that means in terms of this PR; and I guess that review later never came. That's unfortunate, but I'll try and be more patient...

**Comment by FireController1847** [comment]
@LexManos I was told we're waiting on merging this due to 'event-based exp management.' I feel as if this is a separate issue that would require much more reworking than what this PR is trying to accomplish; this PR was made to re-implement the missing XP events... is there any status on what's going to happen to it? I need to know so I know whether or not I should wait to update my mod for this PR or I should continue with my old methods of monitoring XP over the tick.

**Comment by LexManos** [comment]
[Edit] My inital response was from the phone, and was for a different PR.. However I still dont like just firing events for every case.

**Comment by FireController1847** [comment]
Thanks for merging this. I totally understand your concerns and I'd love to help but I have no idea how forge's source code works, and this PR was pretty much the limit of my knowledge. I'm looking forward to seeing any other implementations you guys do for XP though!

**Pull request merged**